File: Text Information/Text Analysis (Singleton).cpp
/**
 * The following program is implemented with the design pattern Singleton,
 *  which takes a file with the extension .txt and analyzes your text.
 **/

#include <iostream>
#include <fstream>
#include <map>

class FileManager {
private:
    static FileManager* instance;
    FileManager() {}

public:
    static FileManager* getInstance() {
        if (instance == nullptr) {
            instance = new FileManager();
        }
        return instance;
    }
    void analyzeFile(const std::string& fileName) {
        try {
            std::ifstream file(fileName);
            if (!file.is_open()) {
                throw std::runtime_error("File not found!");
            }
            
            std::string line;
            int lineCount = 0;
            int wordCount = 0;
            int characterCount = 0;
            std::map<char, int> characterCountMap;
            
            while (std::getline(file, line)) {
                lineCount++;
                // wordCount += std ::count_if(line.begin(), line.end(), [](char c) {return std::isspace(c); }) + 1;
                characterCount += line.size();
                std::istringstream iss(line);
                std::string word;
                while (iss >> word) {
                    wordCount++;
                }
                characterCount += line.size();
                for (char c : line) {
                    if (characterCountMap.find(c) != characterCountMap.end()) {
                        characterCountMap[c]++;
                    } else {
                        characterCountMap[c] = 1;
                    }
                }
            }

            std::cout << "---------------------------------------------------------\n";
            std::cout << "  # File Analysis Summary\n";
            std::cout << "  ## File Name: " << fileName << "\n";
            std::cout << "  ### Line Count: " << lineCount << "\n";
            std::cout << "  #### Word Count: " << wordCount << "\n";
            std::cout << "  ##### Character Count: " << characterCount << "\n";
            std::cout << "---------------------------------------------------------\n";
            std::cout << "  Character Frequencies:\n";
            for (const auto& entry : characterCountMap) {
                std::cout << "    ==> " << entry.first << ": " << entry.second << "\n";
            }
            std::cout << "---------------------------------------------------------\n";
            file.close();
        } catch (const std::exception& e) {
            std::cerr << "Error: " << e.what() << "\n";
        }
    }
};

FileManager* FileManager::instance = nullptr;
int main() {
    FileManager* fileManager = FileManager::getInstance();
    std::string filePath;
    std::cout << "\n---------------------------------------------------------\n";
    std::cout << "  Enter the file path: ";
    std::getline(std::cin, filePath);
    fileManager->analyzeFile(filePath);
    return 0;
}

File: Text Information/Text Analysis (Singleton).py
# The following program is implemented with the design pattern Singleton,
#  which takes a file with the extension .txt and analyzes your text.

class FileManager:
    __instance = None

    @staticmethod
    def getInstance():
        if FileManager.__instance is None:
            FileManager()
        return FileManager.__instance
    def init(self):
        if FileManager.__instance is not None:
            raise Exception("  ( ! This class is a singleton ! )  ")
        else:
            FileManager.__instance = self
    def analyzezFile(self, fileName):
        try:
            with open(fileName, 'r') as file:
                lines = file.readlines()
                lineCount = len(lines)
                wordCount = 0
                characterCount = 0
                characterCountMap = {}                      # dictionary to store character frequencies

                for  line in lines:
                    words = line.split()
                    wordCount += len(words)
                    characterCount += len(line)

                    for  char in line:
                        if char in characterCountMap:
                            characterCountMap[char] += 1
                        else:
                            characterCountMap[char] = 1
                
            print("---------------------------------------------------------")
            print("  # File Analysis Summary")
            print("  ## File Name:", fileName)
            print("  ### Line Count:", lineCount)
            print("  #### Word Count:", wordCount)
            print("  ##### Character Count:", characterCount)
            print("---------------------------------------------------------")
            print("  Character Frequencies:")
            for char, freq in sorted(characterCountMap.items()):
                print("    ==> {}: {}".format(char, freq))
            print("---------------------------------------------------------\n")

        except FileNotFoundError:
            print("  ( ! File not found ! )  ")

if __name__ == '__main__':
    file_Manager = FileManager.getInstance()
    file_Manager = FileManager()
    filePath = input("\n---------------------------------------------------------\n  Enter the file path: ")
    file_Manager.analyzezFile(filePath)